### PR TITLE
Nakamoto Inventory State Machine

### LINKED ISSUE #4062: [Nakamoto] Block inventory sync network state machine
There needs to be a Nakamoto-specific network state machine for synchronizing block inventories.  Per the SIP, the inventory bit-vectors exchanged are framed by reward cycle, and there exists a 1-bit in the ith slot in the vector if there was a tenure-change in that sortition, and a 0-bit if not.  As before, nodes need to synchronize the entirety of their neighbors' block inventory vectors on boot-up, and then only need to track the last two reward cycles' inventories when out of IBD.  Unlike before, there is no need to separately synchronize the PoX bit vector since Nakamoto cannot fork.

### PR DESCRIPTION
**Pull request by jcnelson**
Closes #4062.

This PR implements the Nakamoto tenure inventory state machine, and updates the top-level network work state machine to drive it.

The Stacks blockchain differs from most blockchains in that all peers already know the list of block hashes from the same source of truth -- the sortition history in the Bitcoin blockchain.  Instead of exchanging lists of block hashes, nodes exchange _inventory bitvectors_ to succinctly describe which blocks (or tenures) they have within a given reward cycle's worth of sortitions.  The current system performs this in two steps:  nodes first exchange PoX bitvectors describing which PoX anchor blocks they have, and from there, they exchange reward cycle inventories for each reward cycle in which they agree on the state of the PoX anchor block.

The Nakamoto inventory state machine is simpler, since we assume that all nodes always have all PoX anchor blocks.  Nodes simply request and reply tenure inventories for each reward cycle.

This PR contains some refactoring work in epoch 2.x code paths.  To facilitate reviews, I've annotated where this is.

### CODE REVIEW

File: stackslib/src/net/inv/epoch2x.rs
Comment (jcnelson) [inline stackslib/src/net/inv/epoch2x.rs:2629]: This used to live in `src/net/p2p.rs`.

File: stackslib/src/net/inv/epoch2x.rs
Comment (jcnelson) [inline stackslib/src/net/inv/epoch2x.rs:2737]: This used to be part of `do_network_work()` in `src/net/p2p.rs`.

File: stackslib/src/net/inv/epoch2x.rs
Comment (jcnelson) [inline stackslib/src/net/inv/epoch2x.rs:2662]: This used to be part of `do_network_work()` in `src/net/p2p.rs`.

File: stackslib/src/net/inv/nakamoto.rs
Comment (kantai) [inline stackslib/src/net/inv/nakamoto.rs:521]: Is this method used anywhere or is it only used in tests?

File: stackslib/src/net/inv/nakamoto.rs
Comment (kantai) [inline stackslib/src/net/inv/nakamoto.rs:561]:
```suggestion
        for rc in (highest_rc+1)..=tip_rc {
```

File: stackslib/src/net/inv/nakamoto.rs
Comment (kantai) [inline stackslib/src/net/inv/nakamoto.rs:565]: Should this error instead? This would make the inventory sync somewhat incoherent, right?

File: stackslib/src/net/inv/nakamoto.rs
Comment (kantai) [inline stackslib/src/net/inv/nakamoto.rs:662]:
```suggestion
            let Some(getnakamotoinv) = self.make_getnakamotoinv(inv_rc) else {
                warn!("{:?}: failed to make nakamoto inv, could not find consensus hash for reward cycle.";
                           "reward_cycle" => inv_rc);
                continue;
            };
            if let Err(e) = self.comms.neighbor_send(network, &naddr, getnakamotoinv) {
                    warn!("{:?}: failed to send GetNakamotoInv", network.get_local_peer();
                          "peer" => ?naddr,
                          "error" => ?e
                    );
            }
```

File: stackslib/src/net/p2p.rs
Comment (kantai) [inline stackslib/src/net/p2p.rs:3878]: Should this be explicitly the `Epoch30` boundary? Presumably, a hypothetical `Epoch31` wouldn't want to do Epoch 2.x work during `Epoch30`

File: stackslib/src/net/p2p.rs
Comment (kantai) [inline stackslib/src/net/p2p.rs:3936]:
```suggestion
            debug!(
                "{:?}: network work state is {:?}",
                self.get_local_peer(),
                &self.nakamoto_work_state;
                "learned_new_blocks?" => learned
            );
```

**Review by kantai — COMMENTED** [review]
This LGTM, have some superficial comments throughout, but a few questions that we should clear up before approval.

File: testnet/stacks-node/src/tests/neon_integrations.rs
Comment (jcnelson) [reply testnet/stacks-node/src/tests/neon_integrations.rs:9510]: There's a race condition in `next` at least (and probably also `develop` and `master`) whereby the burnchain block downloader thread races the miner thread, such that a block-commit generated by the miner thread is not mined in the target block.  This, in turn, causes a sortition to be missed, which in turn causes this assertion to fail since the Stacks height would have advanced by 4 blocks instead of 5.  However, the nature of this test is to verify that a block with a problematic transaction gets mined (which happens below); it doesn't matter if it takes 4 or 5 attempts.  This check was needlessly strict.

I fixed it here because I earlier approved a PR to `next` where this test was failing for the same reason, and I told the author not to worry about it since it wasn't blocking the merge (and the code changes were unrelated).

File: stackslib/src/net/inv/nakamoto.rs
Comment (jbencin) [reply stackslib/src/net/inv/nakamoto.rs:521]: Slightly clearer way to write this logic:

```suggestion
    pub fn highest_reward_cycle(&self) -> u64 {
        self.inventories.iter()
            .map(|(_, inv)| inv.highest_reward_cycle())
            .max()
            .unwrap_or(0)
    }
```

File: stackslib/src/net/connection.rs
Comment (jbencin) [reply stackslib/src/net/connection.rs:414]: +1, `rustdoc` comments on struct fields really help the newer developers like myself

File: stackslib/src/net/p2p.rs
Comment (jbencin) [inline stackslib/src/net/p2p.rs:1751]:
```suggestion
                let pubkh = self.get_p2p_convo(event_id)
                    .and_then(|convo| convo.get_public_key_hash())
                    .unwrap_or(Hash160([0x00; 20]));
```

File: stackslib/src/net/p2p.rs
Comment (jbencin) [inline stackslib/src/net/p2p.rs:3877]:
```suggestion
            // not begin on a reward cycle boundary.
```

File: stackslib/src/net/tests/inv/nakamoto.rs
Comment (jbencin) [inline stackslib/src/net/tests/inv/nakamoto.rs:471]: This looks dangerous. As Aaron mentioned earlier, can we use either `stacks_common::BitVec` or the `bitvec` crate (it's already in Cargo.lock)

**Review by jbencin — COMMENTED** [review]
Things to address before I'd approve:
- Some comments need to be fixed
- The manual `BitVec` implementation in `check_inv_state()` looks risky. Can we use an existing and tested implementation?

Other than that, I left some style/readability suggestions, but I won't insist on making those changes

File: stackslib/src/net/inv/epoch2x.rs
Comment (jcnelson) [reply stackslib/src/net/inv/epoch2x.rs:2679]: Sure; this was always in the epoch 2.x state machine, but was simply moved over to a new file.

File: stackslib/src/net/inv/epoch2x.rs
Comment (jcnelson) [reply stackslib/src/net/inv/epoch2x.rs:2683]: This was also just lifted directly from what was running in epoch 2.x

File: stackslib/src/net/inv/epoch2x.rs
Comment (jcnelson) [reply stackslib/src/net/inv/epoch2x.rs:2756]: Sure

File: stackslib/src/net/inv/nakamoto.rs
Comment (jcnelson) [reply stackslib/src/net/inv/nakamoto.rs:323]: Your `BitVec` merged after I wrote this.

File: stackslib/src/net/inv/epoch2x.rs
Comment (jcnelson) [reply stackslib/src/net/inv/epoch2x.rs:2747]: Sure; again, this is just what was already there.  Only mentioning it because I'm trying to save you and @jbencin the time of having to review soon-to-be-legacy code that's been in production for years already ;)

File: stackslib/src/net/inv/nakamoto.rs
Comment (jcnelson) [reply stackslib/src/net/inv/nakamoto.rs:343]: 1. Yes, this is buggy; will fix.  `self.highest_sortition` is the number of sortitions represented by this state machine.  It should be equal to (reward_cycle_len * highest_queried_rc + bitlen_of_highest_queried_rc)

2. The bitvectors themselves are not merged because `tenure_inv` represents the _latest_ knowledge of blocks for the given reward cycle.  If a bit in `tenure_inv` has flipped from 1 to 0 relative to `cur_inv`, or if `tenure_inv` is shorter than `cur_inv`, then that's still pertinent information, since the block downloader ultimately consumes this information to determine which tenures to fetch from this peer.  If a tenure somehow became unavailable on the remote peer, then the block downloader will need to know that so it won't go and request it.

File: stackslib/src/net/inv/nakamoto.rs
Comment (jcnelson) [reply stackslib/src/net/inv/nakamoto.rs:521]: It will be needed by the block downloader.

File: stackslib/src/net/inv/nakamoto.rs
Comment (jcnelson) [reply stackslib/src/net/inv/nakamoto.rs:561]: Sure

File: stackslib/src/net/inv/nakamoto.rs
Comment (jcnelson) [reply stackslib/src/net/inv/nakamoto.rs:565]: Sure

File: stackslib/src/net/inv/nakamoto.rs
Comment (jcnelson) [reply stackslib/src/net/inv/nakamoto.rs:662]: Sure

File: stackslib/src/net/p2p.rs
Comment (jcnelson) [reply stackslib/src/net/p2p.rs:3878]: Yes; I'll constrain this check

File: stackslib/src/net/p2p.rs
Comment (jcnelson) [reply stackslib/src/net/p2p.rs:3936]: Sure

File: stackslib/src/net/inv/nakamoto.rs
Comment (jcnelson) [reply stackslib/src/net/inv/nakamoto.rs:561]: Actually, this is necessary -- tests fail with out it, since we need to populate the consensus hash for `highest_rc` on the first pass.

Comment (kantai) [reply stackslib/src/net/inv/nakamoto.rs:561]: Is that because `highest_rc = 0` could mean either that the 0th reward cycle is the highest one processed or that no reward cycle has been processed?

File: stacks-common/src/bitvec.rs
Comment (jbencin) [inline stacks-common/src/bitvec.rs:167]: Shouldn't this function be:

```rust
self.data.clear();
self.len = 0;
```

File: stackslib/src/net/codec.rs
Comment (jbencin) [inline stackslib/src/net/codec.rs:279]: In general it's preferable to use slices rather than `&Vec` for function parameters

```suggestion
    pub fn compress_bools(bits: &[bool]) -> Vec<u8> {
```

Also, this function looks like it's reimplementing `BitVec::try_from::<bool>()`

**Review by jbencin — APPROVED** [review]
Looks good, just one minor comment

File: stackslib/src/net/inv/nakamoto.rs
Comment (jferrant) [inline stackslib/src/net/inv/nakamoto.rs:355]: Just a nit: we are not particularly consistent with our +1 behaviour. Would we expect this to wrapping add? saturating add? I assume wrapping add. But I tend to avoid straight += 1

**Review by jferrant — APPROVED** [review]
Don't think my review is particularly helpful on this as I don't entirely understand what is happening throughout, but LGTM

File: stackslib/src/net/codec.rs
Comment (jcnelson) [reply stackslib/src/net/codec.rs:279]: This is part of a legacy code path that's live in production, which I'm _very_ keen not to touch

File: stackslib/src/net/inv/nakamoto.rs
Comment (jcnelson) [reply stackslib/src/net/inv/nakamoto.rs:355]: Thanks! I'll make this a saturating add

Comment (jbencin) [reply stackslib/src/net/inv/nakamoto.rs]: Just FYI you can wrap any integer type with `std::num::Saturating` ([link](https://doc.rust-lang.org/std/num/struct.Saturating.html)) to force consistent use of saturating arithmetic operations

File: stackslib/src/net/inv/epoch2x.rs
Comment (kantai) [reply stackslib/src/net/inv/epoch2x.rs]: I'm aware of that, but I see no reason not to apply some simple cleanup during the move.

File: stackslib/src/net/inv/epoch2x.rs
Comment (kantai) [reply stackslib/src/net/inv/epoch2x.rs]: Yes, but I think it's a good opportunity to do simple cleanup as long as the behaviors are preserved.

**Review by kantai — APPROVED** [review]
LGTM!

**Pull request merged**
